File: spikeforest_widgets/widgets/DirectoryView/DirectoryView.js
import React, { Component } from 'react';
import { PythonInterface } from 'reactopya';
import BrowserTree, { TreeData } from './BrowserTree';
const config = require('./DirectoryView.json');

export default class DirectoryView extends Component {
    static title = 'View a directory on a local or remote machine'
    static reactopyaConfig = config
    constructor(props) {
        super(props);
        this.state = {
            // javascript state
            path: null,
            
            // python state
            dir_content: {},
            status: '',
            status_message: ''
        }
        this.treeData = new TreeData;
    }
    componentDidMount() {
        this.pythonInterface = new PythonInterface(this, config);
        this.pythonInterface.start();
        // Use this.pythonInterface.setState(...) to pass data to the python backend
        this.pythonInterface.setState({
            path: this.props.path
        });
    }
    componentWillUnmount() {
        this.pythonInterface.stop();
    }
    render() {
        console.log('---', this.state.dir);
        return (
            <RespectStatus {...this.state}>
                <BrowserTree
                    treeData={this.treeData}
                    dirContent={this.state.dir_content}
                    basePath={this.props.path}
                />
            </RespectStatus>
        )
    }
}

class RespectStatus extends Component {
    state = {}
    render() {
        switch (this.props.status) {
            case 'started':
                return <div>Started: {this.props.status_message}</div>
            case 'running':
                return <div>{this.props.status_message}</div>
            case 'error':
                return <div>Error: {this.props.status_message}</div>
            case 'finished':
                return this.props.children;
            default:
                return <div>Unknown status: {this.props.status}</div>
        }
    }
}

File: spikeforest_widgets/widgets/DirectoryView/DirectoryView.py
import kachery as ka

class DirectoryView:
    def __init__(self):
        super().__init__()

    def javascript_state_changed(self, prev_state, state):
        self._set_status('running', 'Running DirectoryView')

        path = state.get('path')
        dir_content = ka.read_dir(path, recursive=True)

        self._set_state(dir_content=dir_content)

        self._set_status('finished', 'Finished loading DirectoryView')

    def on_message(self, msg):
        # process custom messages from JavaScript here
        # In .js file, use this.pythonInterface.sendMessage({...})
        pass
    
    # Send a custom message to JavaScript side
    # In .js file, use this.pythonInterface.onMessage((msg) => {...})
    def _send_message(self, msg):
        self.send_message(msg)

    # Set the python state
    def _set_state(self, **kwargs):
        self.set_state(kwargs)
    
    # Set error status with a message
    def _set_error(self, error_message):
        self._set_status('error', error_message)
    
    # Set status and a status message. Use running', 'finished', 'error'
    def _set_status(self, status, status_message=''):
        self._set_state(status=status, status_message=status_message)